### DISCUSSION TITLE
Exclude path from matching in redirects/rewrites/headers

### DISCUSSION THREAD

**Discussion by harish-14** [vercel/next.js#34323] [Help]
Hi all!

I am looking at Next.js headers https://nextjs.org/docs/api-reference/next.config.js/headers.

I want to apply security headers for all the except of /abc

I had tried "`(^(?!.*abc).*)`" regex assigning to source key.

Here are my attempts:

Here I want to add a security header, but security header not to applied to abc route:

1.
```
return [
              {
                source: "/(^(?!.*abc).*)",
                headers: securityHeaders,
              }
]
```

2. 
```
return [
              {
                      source: '/(.*)',
                      headers: securityHeaders,
               },
               {
                      // Apply these headers to all routes except for abc route in your application.
                      source: '/abc',
                      headers: securityHeadersABC,
                },
]
```

Any suggestions?

Took a reference from [#24255](https://github.com/vercel/next.js/discussions/24255), but didn't help

**Comment by 2stash** [comment]
Good morning! Have you tried applying the security rules first, then below that having a header for source='/abc'.

Per the docs "If two headers match the same path and set the same header key, the last header key will override the first. "

So theoretically you can just override the abc if you put its logic last in your header config.

**Reply by harish-14** [reply]
Hi @2stash ,
Thank you for your participation in the discussion, realised my regex was wrong 

```
{
        // Apply these headers to all routes except for abc route in your application.
        source: '/((?!abc).*)',
        headers: securityHeaders,
 },
```

this works for me now!

**Reply by 2stash** [reply]
Nice! Yeah I was going to say I'm not trying to read the regex as it hurts my brain lol

**Comment by mdstroebel** [comment]
I am attempting to exclude the `api` folder from my `rewrites` when hosted on Vercel.

```
async rewrites() {
    return {
      beforeFiles: [
        {
          source: '/:host(^(?!.*\/api\/).*$)/:path*', // match any path EXCEPT /api/ paths
          has: [
            {
              type: 'host',
              value: 'subdomain.domain.com', // an example subdomain
            },
          ],
          destination: '/specificFolder/:host/:path*',
        },
      ]
    }
  }
```

I am doing this because I want to have specific subdomains go to specific routes in my pages folder.

e.g. I have `pages/subdomainTest` as a folder of routes and also `pages/subdomainBetty` as a folder of routes. I want the different subdomains to default to different folders.

The above works, but the `api` rewrites always fail because of it... the `api` routes must just use the standard domain and not be rewritten by this rule.

**Comment by mdstroebel** [comment]
Ok, so I seem to have figured out a work-around for this... for some reason I cannot get the regex-path-matching to work. This is my workaround to ensure that the subdomain maps to a new default folder, but maintains the `api` mapping path:

```
async rewrites() {
    return {
      beforeFiles: [
        {
          source: '/:path*',
          has: [
            {
              type: 'host',
              value: 'subdomain1.domain.com',
            },
          ],
          destination: '/subdomain1/:path*',
        },
        {
          source: '/subdomain1/api/:path*', // ensure the API paths are mapped correctly
          has: [
            {
              type: 'host',
              value: 'subdomain1.domain.com',
            },
          ],
          destination: '/api/:path*',
        }
      ]
    }
  }
```

**Comment by kunaljain0212** [comment]
I have a similar usecase where in I need to setup a rewrite for my CMS which is hosted on a separate URL. For example let us assume the CMS is hosted on ``domain1.com`` and I have my main platform deployed to Vercel on some domain say main.com. What I want is to map every path of ``domain1.com`` to ``/r/path`` of main website. That is: when a user tries to access ``main.com/r/admin`` the rewrite should happen to ``domain1.com/admin`` and so on for every possible path of ``domain1.com``. Can someone please help me out with this?
